timeutil/schedule.go

```diff
@@ -82,40 +108,285 @@ func ParseTime(s string) (t TimeOfDay, err error) {
+		// day span, eg. mon-fri
+		return false
+	case wdEnd < wdStart && t.Weekday() < wdStart && t.Weekday()+7 > wdEnd+7:
+		// day span that wraps around, eg. fri-mon,
+		// since time.Weekday values go from 0-6, add 7
+		// (week) at the end to get a continuous range
+		return false
+	}
+	return true
+}
+
+// TimeSpan is a time span within a single day (or this and the next day).
+// TimeSpan may wrap around, eg. 23:00-1:00 representing a span from 11pm to 1am
+// the next day.
+type TimeSpan struct {
 	Start TimeOfDay
 	End   TimeOfDay
+	// indicates if the span is split into smaller subspans. Use Subspan()
+	// to generate a slice of TimeSpan using the current TimeSpan as a base
+	Split uint
+	// indicates if the events are randomly spread between Start and End
+	Spread bool
+}
+
+func (ts TimeSpan) String() string {
+	sep := "-"
+	if ts.Spread {
+		sep = "~"
+	}
+	if !ts.End.IsZero() && ts.End != ts.Start {
+		s := ts.Start.String() + sep + ts.End.String()
+		if ts.Split > 0 {
+			s += "/" + strconv.FormatUint(uint64(ts.Split), 10)
+		}
+		return s
+	}
+	return ts.Start.String()
+}
+
+// MakeTime generates a start and end times from ts using t as a base. Returned
+// end time is automatically shifted to the next day if End is before Start
+func (ts TimeSpan) MakeTime(t time.Time) (start time.Time, end time.Time) {
+	a := ts.Start.MakeTime(t)
+	b := a
+	if !ts.End.IsZero() && ts.End != ts.Start {
+		b = ts.End.MakeTime(t)
+
+		// 23:00-1:00
+		if b.Before(a) {
+			b = b.Add(24 * time.Hour)
+		}
+	}
+	return a, b
+}
+
+// Subspans returns a slice of TimeSpan generated from ts by splitting the time
+// between ts.Start and ts.End into ts.Split equal spans.
+func (ts TimeSpan) Subspans() []TimeSpan {
+	if ts.Split == 0 || ts.Split == 1 || ts.End == ts.Start {
+		return []TimeSpan{ts}
+	}
+
+	span := ts.End.Sub(ts.Start)
+	step := span / time.Duration(ts.Split)
+
+	var spans []TimeSpan
+	for i := uint(0); i < ts.Split; i++ {
+		start := ts.Start.Add(time.Duration(i) * step)
+		spans = append(spans, TimeSpan{
+			Start:  start,
+			End:    start.Add(step),
+			Split:  0, // no more subspans
+			Spread: ts.Spread,
+		})
+	}
+	return spans
+}
 
-	Weekday string
+// Schedule represents a single schedule
+type Schedule struct {
+	Week  []WeekSpan
+	Times []TimeSpan
 }
 
 func (sched *Schedule) String() string {
-	return fmt.Sprintf("%s-%s", sched.Start, sched.End)
+	buf := &bytes.Buffer{}
+
+	for i, span := range sched.Week {
+		if i > 0 {
+			fmt.Fprint(buf, ",")
+		}
+		fmt.Fprintf(buf, "%s", span)
+	}
+
+	if len(sched.Week) > 0 && len(sched.Times) > 0 {
+		fmt.Fprint(buf, ",")
+	}
+
+	for i, span := range sched.Times {
+		if i > 0 {
+			fmt.Fprintf(buf, ",")
+		}
+		fmt.Fprintf(buf, "%s", span)
+	}
+	return buf.String()
 }
 
-func (sched *Schedule) Next(last time.Time) (start, end time.Time) {
+func (sched *Schedule) weekSpans() []WeekSpan {
+	if len(sched.Week) == 0 {
+		return []WeekSpan{{}}
```

Comment:
Alternatively, if the algorithm looks simpler with this having a default span, we need to be explicit and cook a proper span instead of trusting zero values.